src/Elastic.Apm.SqlClient/SqlEventListener.cs

```diff
@@ -0,0 +1,139 @@
+using Elastic.Apm.Api;
+using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
+
+namespace Elastic.Apm.SqlClient
+{
+	internal class SqlEventListener : EventListener
+	{
+		private readonly ApmAgent _apmAgent;
+		private readonly IApmLogger _logger;
+
+		private readonly ConcurrentDictionary<int, (Span Span, long Start)> _spans = new ConcurrentDictionary<int, (Span, long)>();
+
+		public SqlEventListener(IApmAgent apmAgent)
+		{
+			_apmAgent = (ApmAgent)apmAgent;
+			_logger = _apmAgent.Logger.Scoped(nameof(SqlEventListener));
+		}
+
+		protected override void OnEventSourceCreated(EventSource eventSource)
+		{
+			if (eventSource != null && eventSource.Name == "Microsoft-AdoNet-SystemData")
+				EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)1);
+
+			base.OnEventSourceCreated(eventSource);
+		}
+
+		private const int BeginExecuteEventId = 1;
+		private const int EndExecuteId = 2;
+
+		protected override void OnEventWritten(EventWrittenEventArgs eventData)
+		{
+			if (eventData?.Payload == null) return;
+
+			try
+			{
+				switch (eventData.EventId)
+				{
+					case BeginExecuteEventId:
+						ProcessBeginExecute(eventData.Payload);
+						break;
+					case EndExecuteId:
+						ProcessEndExecute(eventData.Payload);
+						break;
+				}
+			}
+			catch (Exception ex)
+			{
+				_logger?.Error()?.LogException(ex, "Error has occurred during handle event from SqlClient. EventData: {EventData}", eventData);
+			}
+		}
+
+		private void ProcessBeginExecute(IReadOnlyList<object> payload)
+		{
+			if (payload.Count != 4)
+			{
+				_logger?.Debug()
+					?.Log("BeginExecute event has {PayloadCount} payload items instead of 4. Event processing is skipped.", payload.Count);
+				return;
+			}
+
+			var id = Convert.ToInt32(payload[0]);
+			var dataSource = Convert.ToString(payload[1]);
+			var database = Convert.ToString(payload[2]);
+			var commandText = Convert.ToString(payload[3]);
+			var start = Stopwatch.GetTimestamp();
+
+			_logger?.Trace()
+				?.Log("Process BeginExecute event. Id: {Id}. Data source: {DataSource}. Database: {Database}. CommandText: {CommandText}.", id,
+					dataSource, database, commandText);
+
+			var spanName = !string.IsNullOrWhiteSpace(commandText)
+				? commandText.Replace(Environment.NewLine, "")
+				: database;
+
+			var span = (Span)ExecutionSegmentCommon.GetCurrentExecutionSegment(_apmAgent)?.StartSpan(spanName, ApiConstants.TypeDb);
+			if (span == null) return;
+
+			if (_spans.TryAdd(id, (span, start)))
+			{
+				span.Context.Db = new Database
+				{
+					Statement = !string.IsNullOrWhiteSpace(commandText)
+						? commandText.Replace(Environment.NewLine, "")
+						: string.Empty,
+					Instance = database,
+					Type = Database.TypeSql
+				};
+
+				span.Context.Destination = _apmAgent.TracerInternal.DbSpanCommon.GetDestination($"Data Source={dataSource}", false, null);
+
+				// System.Data.SQLite and Microsoft.Data.Sqlite don't spread events via EventSource, however,
```

Comment:
I'm not sure, but given the evensource name is `Microsoft-AdoNet-SystemData` I'd be very surprised if any other provider would trigger this code. I assume this is tied to the assembly that hosts `SqlClient` by Microsoft, therefore I think subtype mssql would be ok here. 

Of course I could be wrong - nevertheless I'd just set it here.

Reply:
I agree with you and this is why I left a comment: to decide what to do it with on review.
I don't know why I didn't point to it in the PR description. BTW, set span subtype to `mssql`